src/Pact/Analyze/Types/Model.hs

```diff
@@ -27,49 +31,114 @@ data Arg = Arg
+  deriving (Num, Enum, Show, Ord, Eq)
+
+data Recoverability
+  -- The path upon which to resume inclusion of events. An alternative here
+  -- would be to have more edges in the subgraphs formed by @enforce-one@ --
+  -- have vertices not just for each case, but additionally one after each
+  -- recoverable assert/auth as well, to connect from right after the
+  -- assert/auth to the next case. One would need to be careful here not to try
+  -- to make two edges between a pair of vertices when an assert/auth in a
+  -- preceeding case is right before the next case. We need to explicitly talk
+  -- about the path upon which to resume execution because it's not necessarily
+  -- the next vertex -- there could be more subpaths before where we should
+  -- resume due to nested conditionals or @enforce-one@s.
+  = Recoverable { _resumptionPath :: TagId }
+  | Unrecoverable
+  deriving (Eq, Show)
+
+-- For determining resumption paths for nested @enforce-one@s:
+instance Semigroup Recoverability where
+  --      outer <> inner
+  Recoverable _ <> Recoverable y = Recoverable y
+  Recoverable x <> Unrecoverable = Recoverable x
+  Unrecoverable <> nested        = nested
+
+instance Monoid Recoverability where
+  mempty = Unrecoverable
+  mappend = (<>)
+
+data TraceEvent
+  = TraceRead (Located (TagId, Schema))
+  | TraceWrite (Located (TagId, Schema))
+  | TraceAssert Recoverability (Located TagId)
+  | TraceAuth Recoverability (Located TagId)
+  | TraceBind (Located (VarId, Text, EType))
+  | TraceSubpathStart TagId
+  deriving (Eq, Show)
+
+data ExecutionGraph
+  = ExecutionGraph
+    { _egInitialVertex :: Vertex
+    , _egRootPath      :: TagId
+    , _egGraph         :: Alga.Graph Vertex
+    , _egEdgeEvents    :: Map Edge [TraceEvent]
+    , _egPathEdges     :: Map TagId [Edge]
+    }
+  deriving (Eq, Show)
+
+data Concreteness
+  = Concrete
+  | Symbolic
+
+data ModelTags (c :: Concreteness)
   = ModelTags
-    { _mtVars   :: Map VarId (Located (Text, TVal))
+    { _mtVars    :: Map VarId (Located (Text, TVal))
     -- ^ each intermediate variable binding
-    , _mtReads  :: Map TagId (Located (S RowKey, Object))
-    -- ^ one per each read, in traversal order
-    , _mtWrites :: Map TagId (Located (S RowKey, Object))
-    -- ^ one per each write, in traversal order
-    , _mtAuths  :: Map TagId (Located (SBV Bool))
-    -- ^ one per each enforce/auth check, in traversal order. note that this
-    -- includes all (enforce ks) and (enforce-keyset "ks") calls.
-    , _mtResult :: Located TVal
+    , _mtReads   :: Map TagId (Located (S RowKey, Object))
+    -- ^ one per each read
+    , _mtWrites  :: Map TagId (Located (S RowKey, Object))
+    -- ^ one per each write
+    , _mtAsserts :: Map TagId (Located (SBV Bool))
+    -- ^ one per non-keyset enforcement
+    , _mtAuths   :: Map TagId (Located (S KeySet, SBV Bool))
+    -- ^ one per each enforce/auth check. note that this includes all
+    -- @(enforce ks)@ and @(enforce-keyset "ks")@ calls.
```

Comment:
Can you do `enforce ks`?

Reply:
Good call. I meant `(enforce-keyset "ks-name")` and `(enforce-keyset ks)`